[wallet] abort when attempting to fund a transaction above -maxtxfee

Sjors opened the pull request:
`FundTransaction` calls `GetMinimumFee` which, when the fee rate is absurdly high, quietly reduces the fee to `-maxtxfee`.

Becaue an absurdly high fee rate is usually the result of a fat finger, aborting seems safer behavior.

Before:
```
bitcoin-cli walletcreatefundedpsbt '[]' '[{"tb1q...": 0.01}]' 0 '{"feeRate": 10}' true
{
  "psbt": "cHNidP8...gAA=",
  "fee": 0.10000000,
  "changepos": 1
}

```

After:
```
bitcoin-cli walletcreatefundedpsbt '[]' '[{"tb1q...": 0.01}]' 0 '{"feeRate": 10}' true
error code: -25
error message:
Fee exceeds maximum configured by -maxtxfee
```

QT still checks the max fee rate as expected:
<img width="566" alt="Schermafbeelding 2019-06-20 om 19 52 00" src="https://user-images.githubusercontent.com/10217/59888424-a2aa7100-9395-11e9-8ae6-8a3c1f7de585.png">

promag commented:
Also affects `fundrawtransaction`. Despite the fact that these are low level calls and the client can also discard high fees, I guess it's ok to be on the safe side.

Deserves a small release note like

> walletcreatefundedpsbt and fundrawtransaction now fail to create transactions with a resulting fee greater then -maxtxfee.

Concept ACK.

Sjors commented:
In additional to small transactions with a fat fingered fee rate, there's also an issue with extremely large transactions, which silently get their fee rate reduced. I didn't check if this PR fixes that as well, but I would assume so.

See also #16192 and #16192.

Sjors commented:
Full list of transactions that pay exactly 0.1 BTC in fees: https://blockchair.com/bitcoin/transactions?q=fee(10000000)# (there could be other wallets doing that of course)

practicalswift commented:
Concept ACK: user safety first!

gertjaap commented:
If aborting above `maxtxfee` is the desired behavior, then this PR should be merged in stead of #16192 because it catches that situation too.

achow101 commented:
Are we certain that this does not affect the other methods of transaction creationg (i.e. `CreateTransaction`)? Are there tests for absurdly high fees in regular wallet transactions, and if not, could you add some?

Sjors commented:
One specific case this prevents is when a user sets `feeRate: 1` (thinking it's satoshi per byte). A 1 SegWit input to 1 output transaction is 110 vbytes or 0.11 kilobyte. At 1 BTC per kilobyte the fee for that transaction would be 0.11 BTC which is just above `-maxfee`.

@achow101 `CreateTransaction` is called by:
* `CreateRateBumpTransaction`
* `sendtoaddress` (via `SendMoney`)
* `sendmany`
* `fundrawtransaction` (via `FundTransaction`)

I'll try to add some tests for those. Note that `sendtoaddress` and `sendmany` don't have a `feeRate` argument.

jonasschnelli commented:
1) in coin control, absurd fees would no longer be capt to maxtxfee (which probably is okay, just that we are aware). You can't send from the GUI with fees > maxfee, but you currently can play with absurde fees in coin-control (I don't think this makses sense).

2) There is a comment in walletmodel.cpp that needs overhaul:
```
        // reject absurdly high fee. (This can never happen because the
        // wallet caps the fee at m_default_max_tx_fee. This merely serves as a
        // belt-and-suspenders check)
        if (nFeeRequired > m_wallet->getDefaultMaxTxFee())
            return AbsurdFee;
```

3) verified that the impact of removing the FeeReason::MAXTXFEE (and remove of the silent auto-cap to maxfee) in other places then FundTransaction are non-significant.

Tested ACK 5735eb4749cdd7cddf0303e150493a6381b656f5

meshcollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3200,6 +3200,10 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
     }
+
+    if (fee_out > pwallet->m_default_max_tx_fee) {
+      throw JSONRPCTransactionError(TransactionError::MAX_FEE_EXCEEDED);
```
nit: indentation

meshcollider reviewed (approved):
LGTM 5735eb4749cdd7cddf0303e150493a6381b656f5

Much safer behaviour IMO

Sjors commented:
Added release note and test for `fundrawtransaction`.

kallewoof commented:
Concept ACK

MarcoFalke commented on src/qt/walletmodel.cpp:
```diff
@@ -221,9 +221,7 @@ WalletModel::SendCoinsReturn WalletModel::prepareTransaction(WalletModelTransact
 
-        // reject absurdly high fee. (This can never happen because the
-        // wallet caps the fee at m_default_max_tx_fee. This merely serves as a
-        // belt-and-suspenders check)
```
I don't like removing this belt-and-suspenders check and making it the only check.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3208,6 +3208,10 @@ void FundTransaction(CWallet* const pwallet, CMutableTransaction& tx, CAmount& f
+
+    if (fee_out > pwallet->m_default_max_tx_fee) {
+        throw JSONRPCTransactionError(TransactionError::MAX_FEE_EXCEEDED);
+    }
```
I don't mind having a belt-and-suspenders check in the rpc code (similar to the GUI code), but ultimately the check should happen in the wallet.

In fact the wallet already does this check in a hacky way (`Transaction too large`...). So I'd prefer if you removed the hacky check and replaced it with a check against m_default_max_tx_fee

Sjors replied:
It indeed makes sense for the wallet to do this check. Will look into that.

Sjors replied:
Done

maflcko reviewed (commented):
Concept ACK, some style questions

Sjors commented on doc/release-notes-16257.md:
```diff
@@ -0,0 +1,6 @@
+Wallet changes
```
I think this is equally useful as a master release note.

MarcoFalke replied:
I think we put it in the release notes only of the earliest release that contains the new feature/or bugfix. But not blocking of course.

promag commented:
@jonasschnelli not sure if I understand your review, it's still possible to set a high fee rate that causes the fee capping effectively reducing the fee rate. I think this PR should also change the GUI behaviour, or is there a reason to have them different?

Sjors committed:
[wallet] abort when attempting to fund a transaction above maxtxfee

FundTransaction calls GetMinimumFee which, when the fee rate is absurdly high, quietly reduced the fee to -maxtxfee. Becaue an absurdly high fee rate is usually the result of a fat finger, aborting seems safer behavior.

Sjors commented:
I moved the check from RPC to the wallet. On a related note, I have in a [work in progress branch](https://github.com/Sjors/bitcoin/tree/2019/06/TransactionError) which replaces all `_failReason`  occurrences with `TransactionError`.

promag commented:
> @promag I'd rather not change the GUI messaging here.

@Sjors fair enough.

Is it me or RPC `bumpfee` (thru `feebumper::CreateRateBumpTransaction`) doesn't check for `-maxtxfee`?

laanwj reviewed (approved):
Code review ACK 806b0052c3b45415862f74f20ba5f389e5b673de

MarcoFalke commented:
@promag Looks like it. At the very least we should have a test for this.

Sjors commented:
I made a note to look into `bumpfee` and either add a test or fix the issue (happy to review if someone else does it first).

Jaker9 commented:
3Djkyp7ajw1ykhuy55DNmsp3VBMeWth44h
Help me

practicalswift commented on test/functional/rpc_psbt.py:
```diff
@@ -129,6 +130,15 @@ def run_test(self):
+        assert_greater_than(0.06, res["fee"])
+
+        # feeRate of 10 BTC / KB produces a total fee well above -maxtxfee
+        # previously this was silenty capped at -maxtxfee
```
Post-merge typo: "silenty" :-)

hebasto replied:
Why didn't Travis catch it while running `test/lint/lint-spelling.sh`?

practicalswift replied:
I think it does, but `lint-spelling.sh` intentionally does not break the build in case of new typos. It just prints them to the Travis log for periodic speling correction :-)

fanquake commented:
Being backported in #16414.

Sjors commented:
0.17 backport in #16639

gmaxwell commented:
oy.  this interface takes feerates in btc but most ecosystem stuff takes feerates in satoshis.  :(
